Fix memory leak in 'ProxySQL_Admin::stats___mysql_global' - Closes #4976

JavierJF opened the pull request:
## Description

The resource leak was associated to prepared statements (sqlite3_stmt) used on 'ProxySQL_Admin::stats___mysql_global'.

## Fix

Since the detection of these leaks is tricky, because the resources held by 'SQLite3' never lose their references, a helper type ('stmt_unique_ptr') and a safer variation for 'prepare_v2' have been introduced. This helper type should automatically handle the lifetime of the statements, avoiding these kind of leaks in the future. The non-safe alternative has been flagged as 'deprecated', so it can be easily identified in later refactors or slowly removed during development.

JavierJF committed:
Fix memory leak in 'ProxySQL_Admin::stats___mysql_global' - Closes #4976

The resource leak was associated to prepared statements (sqlite3_stmt).
Since the detection of these leaks is tricky, because the resources held
by 'SQLite3' never lose their references, a helper type
('stmt_unique_ptr') and a safer variation for 'prepare_v2' have been
introduced. This helper type should automatically handle the lifetime of
the statements, avoiding these kind of leaks in the future. The non-safe
alternative has been flagged as 'deprecated', so it can be easily
identified in later refactors or slowly removed during development.

renecannao commented:
Can one of the admins verify this patch?

JavierJF commented on include/sqlite3db.h:
```diff
@@ -191,7 +206,14 @@ class SQLite3DB {
+	 * @brief Prepares a query as an statement in the SQLite3DB.
+	 * @param The query to be prepared as an 'sqlite3_stmt'.
+	 * @return A pair of with shape { err_code, stmt_unique_ptr }.
+	 */
```
The typo will be addressed, the param tag is missing because the definition doesn't have a param name as there is only one param (which is valid). But, I will change it for clarity and change the parameter name too for the same reason.

JavierJF commented on lib/ProxySQL_Admin_Stats.cpp:
```diff
@@ -521,12 +521,17 @@ void ProxySQL_Admin::stats___mysql_global() {
+
+	stmt_unique_ptr u_row_stmt { nullptr };
+	std::tie(rc, u_row_stmt) = statsdb->prepare_v2(q_row_insert.c_str());
 	ASSERT_SQLITE_OK(rc, statsdb);
-	sqlite3_stmt* bulk_stmt = nullptr;
-	rc = statsdb->prepare_v2(q_bulk_insert.c_str(), &bulk_stmt);
+	sqlite3_stmt* row_stmt { u_row_stmt.get() };
```
The default initialization will stay, I don't like brevity in the expense of clarity, it imposes more mental load trying to remember what implicit inits take place (by memory), than seeing an explicit initialization. The `const` make change, I left them like the originals, but a const makes the intentions more clear (now that is also managed).

JavierJF commented on lib/ProxySQL_Admin_Stats.cpp:
```diff
@@ -521,12 +521,17 @@ void ProxySQL_Admin::stats___mysql_global() {
-	rc = statsdb->prepare_v2(q_bulk_insert.c_str(), &bulk_stmt);
+	sqlite3_stmt* row_stmt { u_row_stmt.get() };
+
+	stmt_unique_ptr u_bulk_stmt { nullptr };
+	std::tie(rc, u_bulk_stmt) = statsdb->prepare_v2(q_bulk_insert.c_str());
 	ASSERT_SQLITE_OK(rc, statsdb);
+	sqlite3_stmt* bulk_stmt { u_bulk_stmt.get() };
```
Same as the previous response.

cyberplant commented:
One more thing I noticed.. I left one server with this bug running but without any usage, and I still see memory usage increasing. Could it be because of PMM monitoring?

Also, getting the memory usage takes long time:
```
ProxySQLAdmin >select * from stats.stats_memory_metrics;
+-----------------------------------+----------------+
| Variable_Name                     | Variable_Value |
+-----------------------------------+----------------+
| SQLite3_memory_bytes              | -1508139680    |
| jemalloc_resident                 | 5133131776     |
| jemalloc_active                   | 4991012864     |
| jemalloc_allocated                | 4967472712     |
| jemalloc_mapped                   | 5293178880     |
| jemalloc_metadata                 | 141019480      |
| jemalloc_retained                 | 1739096064     |
| Auth_memory                       | 21596          |
| mysql_query_digest_memory         | 1757056320     |
| mysql_query_rules_memory          | 276            |
| pgsql_query_digest_memory         | 0              |
| pgsql_query_rules_memory          | 274            |
| prepare_statement_metadata_memory | 563810         |
| prepare_statement_backend_memory  | 338720         |
| mysql_firewall_users_table        | 0              |
| mysql_firewall_users_config       | 0              |
| mysql_firewall_rules_table        | 0              |
| mysql_firewall_rules_config       | 329            |
| stack_memory_mysql_threads        | 67108864       |
| stack_memory_admin_threads        | 92274688       |
| stack_memory_cluster_threads      | 41943040       |
+-----------------------------------+----------------+
21 rows in set (1.46 sec)
```

I started looking at this because of some slowness in ProxySQL. After a restart the memory usage goes down and speed returns back to normal.

If it helps you, I can test the new version on this server, at least to see if memory usage goes up without real usage, as it currently does.

JavierJF commented:
Hi @cyberplant,

that's expected, since the leak is in `stats_mysql_global` and since (iirc) PMM constantly queries that table, even if the instance is idle, the leak is going to be exercised and the memory usage will increase. The leak should be fixed in this branch, we already performed testing to verify it, but you can of course test this branch to verify it on your own. Regarding the potential overflow, I will also take care of it soon in this branch.

Thanks, regards,
Javier.

JavierJF committed:
Fix potential overflow for 'sqlite3_status' memory metrics

The previous call have been replaced by 'sqlite3_status64' when querying
'SQLITE_STATUS_MEMORY_USED' to avoid potential integer overflows.

cyberplant commented:
Hey @JavierJF!

I'm also observing high CPU usage on the ProxySQL instance, even when idling. Could this somehow be related to the same issue or it's unrelated?

To clarify: 
I still have one instance running with 3.0.1 without restarting it, but it's not processing production traffic.
```
proxysql 4106365 53.8  2.5 9263868 6753836
```

Memory usage is high, but not sure if CPU usage can also be related to this issue, could it be?

Is there a way to check what's is doing, besides processlist? I checked also with `strace` but didn't spot anything useful.

JavierJF commented:
Hi @cyberplant,

sorry I missed this one. Half a core seems like too much for a idling instance. CPU usage shouldn't be related to this issue, specially since according to the `ps` you shared the instance has plenty of memory available, please correct me otherwise. Said this, I'm not sure if an overhead can be expected in `SQLite3` when the number of prepared statements grows very high without freeing.

To be certain that the issue is gone the simplest thing would be to test this patch in your instance, if CPU usage never increases, my guess would be an overhead in `SQLite3` motivated from the (very) high of non-closed statements. If on the contrary you still experience the CPU overhead, I would suggest opening a different issue and sharing the details of your setup. We could follow up there to understand the reason of the overhead.

Thanks, regards,
Javier.

cyberplant commented:
Excellent, I'll test this and if I see the issue still happening, I'll open a new ticket with some data for investigation.

For testing your PR, does the CI/CD have some package built for me, or I have to build the package by myself from this branch?

JavierJF commented:
Hi @cyberplant,

I can provide you with builds for this PR if you require them, you can write me the details of your OS for the build at javier@proxysql.com. I will share the build for testing privately there.

Regards, Javier.

cyberplant commented:
Just FTR: Had the test build running for some days, and the leak fix looks good. No increase in memory as it was happening before.

Today I put the server in production to check how it goes with production load and still looks good. 👍

JavierJF commented:
Hi @cyberplant,

thanks for testing the build and confirming that CPU increase was likely a side effect.

Regards, Javier.

cyberplant commented:
> Hi @cyberplant,
> 
> thanks for testing the build and confirming that CPU increase was likely a side effect.

Memory usage has been stable now with the test build, but the CPU usage is still increasing, so I would say it's not related to this issue. For example, yesterday the query `select * from stats.stats_memory_metrics;` took 0.01 seconds (sometimes 0.00) and now it takes 0.03 consistently. I will investigate more and open a new issue for this.

Thanks!
